Multithreading support and download videos hosted on YouTube

satyamtg opened the pull request:
This fixes #66 . We now use youtube-dl to download videos for which there's no direct download link.

rgaudin commented:
Hum, you definitely need to elaborate on why that is required. It's an important move that raises additional questions:

* why can't we get the link info directly?
* how is the youtube-dl TED plugin doing ?
* why not just getting the download link from yt-dl instead of using it to download?
* is TED frequently changing this? I don't think so. As we depend heavily on their architecture for the rest (metadata, etc), we might want to just duplicate the yt-dl method and integrate it.

satyamtg commented:
> * why can't we get the link info directly?
Okay. It appears that some videos do not have direct download links (not hosted on TED servers). Their download links section is as follows
```
				"nativeDownloads": null,
				"subtitledDownloads": null,
				"audioDownload": null
```
However, these videos are actually hosted on YouTube and instead have the youtube video ID in a section called external in the JSON data that contains the details. It looks like this -
```
"external": {
					"service": "YouTube",
					"code": "SYB5r3_6vFU",
					"duration": 743.0,
					"start_time": 0.0
				}
```
So, we would actually miss the videos which are hosted on YouTube if we don't take care of this. It's just that all TED videos are not on TED servers and all TED videos are not on YouTube. So we need to handle both.

> * how is the youtube-dl TED plugin doing ?

Actually it works pretty fine (got no errors in all the tests). We just download video from YouTube and still download thumbnail and speaker image from TED servers. Such videos have both a TED video ID and a YouTube video ID. We download in the folder with TED video ID but use the YT video ID to download.

> * why not just getting the download link from yt-dl instead of using it to download?
IMO if we get the link we are not sure what format we're going to download in and YT links do have expiry dates/times. So we need to download as quickly as possible and as far as possible in the same video format that we want the output in (this would save us some conversion time if high quality is requested).



> * is TED frequently changing this? I don't think so. 

No. TED hasn't been frequently changing it. The JSON structure has not changed since the previous release of TED (before the rewrite). 

> As we depend heavily on their architecture for the rest (metadata, etc), we might want to just duplicate the yt-dl method and integrate it.
Am, actually I could not get this very clear. If this means we download all videos from YT instead of TED servers, then actually all of them are not on YT. We cannot do that.

rgaudin commented:
Ah! videos are hosted on youtube… OK that was unexpected and is actually a huge drawback.
I see that the example you mention is not a TED talk but a TEDx one.
That makes sense as historically, TED is only about TED talks and the sister projects have always been hosted elsewhere.

The main problem with this is that youtube has different and more restrictive limitations on usage. 
This results in this scraper being subject to both TED throttling (handled already) and youtube…

I don't think it makes sense to exclude non-TED talks from a topic selection or to add another argument for that. @kelson42 ? That said, looking into this, I realized there's this `videocontext` property that gives the *source* of the video and that might be interesting to include in the UI make-over.

I think this forces us to introduce failure-due-to-429 detections in the Youtube-retrieval code with delays… honestly this is becoming a clusterfuck as we'd have to seggregate YT-downloads into a different queue to accomodate throttling… so **a lot** of refactoring

satyamtg commented:
@rgaudin this actually needs a refactor in the way we download everything. Currently the system we use can be divided as follows -
Step 1 - Get all metadata  (Only subject to TED throttling) (No multithreading)
Step 2 - Get all assets - This contains the following sub-steps
    a. Download video (Subject to TED throttling and youtube throttling) (No multithreading)
    b. Download speaker image (Only subject to TED throttling) (No multithreading)
    c. Download thumbnail (Only subject to TED throttling) (No multithreading)

But, since we want to handle youtube throttling independent of TED throttling, we certainly need to do it in two queues. So, I think it makes sense to have it done in 3 steps instead of 2. As follows -
Step 1 - Get all metadata (we add multithreading in one queue as this step is quite a bit slow + and maybe add user agent faking) - For this step, we can divide the part where we have a list of video pages to scrape data from into batches and download the data in each batch serially.
Step 2 - Download videos (have 2 queues here, run parallel,  one for YT, other for TED) - We can handle all youtube links as done in youtube scraper in batches. Also we may have list of success and failures. For TED also, we introduce multithreading in a similar way as step 1 and have list of success and failures.
Step 3 - Download thumbnails and speaker images for succeeded videos. We can do this in one queue in batches similar to 1.

So, overall, we may divide the download_video_files into 2 parts, download_video and download_images. The download_video would be responsible for creating batches of YT and Ted queues and executing respective batch download functions, or else, downloading serially if concurrency is 1.

satyamtg commented:
I have rebased this branch on the `webp_support` branch and it now has the following changes -
- Introduced a `--threads` argument for specifying how many threads to use at max
- Added support for downloading missing videos which were hosted on YouTube
- Video/Image downloading is parallelized via `download_video_files_parallel()`
- subtitle downloading is parallelized via `download_subtitles_parallel()`
- Added `extract_download_link()` to get the download link or the youtube video ID

Though this can be further improved with better parallelization (by collecting download links together and having a unified parallelized download function), but this seems to give it a significant increase in offlining time without drastically modifying the current working structure.

rgaudin reviewed (changes requested):
Thanks ; scraperlib api may change but there's also a convertion for S3 to fix I believe

satyamtg commented on ted2zim/scraper.py:
```diff
@@ -787,105 +810,130 @@ def download_thumbnail(
+        """ download videos and images parallely """
+
+        self.yt_downloader = YoutubeDownloader(threads=1 if self.threads == 1 else 2)
+        with concurrent.futures.ThreadPoolExecutor(
```
This is because of the following reasons -
1. All videos are not hosted on YouTube. The scraperlib thing only parallelizes (or throttles) the youtube downloading
2. It does everything related to videos - Downloading videos, thumbnails and speaker images, and upload to cache

rgaudin replied:
OK

satyamtg commented on ted2zim/scraper.py:
```diff
@@ -787,105 +810,130 @@ def download_thumbnail(
+
+        # recompress if necessary
+        try:
+            post_process_video(
```
That was passed to skip recompress as we previously used to convert thumbnail in that function also. I have changed to a more sensible thing as this behavior has now been changed.

rgaudin commented on ted2zim/scraper.py:
```diff
@@ -849,7 +849,16 @@ def download_video_files(self, video):
             try:
                 if "https://" not in video_link:
-                    self.yt_downloader.download(video_link, org_video_file_path)
+                    options = (
```
```py
options = (BestWebm if video_format == "webm" else BestMp4).get_options(
    target_dir=video_dir, filepath=pathlib.Path("video.%(ext)s"))
```

satyamtg replied:
Done

rgaudin commented on requirements.txt:
```diff
@@ -1,5 +1,5 @@
 python-dateutil==2.8.1
-zimscraperlib>=1.3.2,<1.4
+zimscraperlib@git+git://github.com/openzim/python_scraperlib@youtube_downloader#egg=zimscraperlib
```
doesn't work now that it's been merged

rgaudin reviewed (changes requested):
Making a release of scraperlib so this can be integrated and rebased before merge. Please fix the inline comment.
